File: examples/node/basic.mjs
/**
 * turbolite example — Node.js (http + koffi)
 *
 * A small HTTP API server backed by turbolite-compressed SQLite.
 *
 * Run:
 *   make lib-bundled
 *   make example-node
 *
 * Then:
 *   curl -X POST localhost:3000/books -d '{"title":"Dune","year":1965}'
 *   curl localhost:3000/books
 */

import koffi from "koffi";
import http from "http";
import { mkdtempSync, rmSync } from "fs";
import { tmpdir } from "os";
import { join, resolve, dirname } from "path";
import { fileURLToPath } from "url";

const __dirname = dirname(fileURLToPath(import.meta.url));

// ── Load turbolite ──────────────────────────────────────────────────

const ext = process.platform === "darwin" ? "dylib" : "so";
const lib = koffi.load(
  resolve(__dirname, `../../target/release/libsqlite_compress_encrypt_vfs.${ext}`)
);

const turbolite_version = lib.func("turbolite_version", "str", []);
const turbolite_last_error = lib.func("turbolite_last_error", "str", []);
const turbolite_register_compressed = lib.func("turbolite_register_compressed", "int", ["str", "str", "int"]);
const turbolite_open = lib.func("turbolite_open", "void*", ["str", "str"]);
const turbolite_exec = lib.func("turbolite_exec", "int", ["void*", "str"]);
const turbolite_query_json = lib.func("turbolite_query_json", "str", ["void*", "str"]);
const turbolite_close = lib.func("turbolite_close", "void", ["void*"]);

// ── Set up database ─────────────────────────────────────────────────

const dataDir = mkdtempSync(join(tmpdir(), "turbolite-example-"));
turbolite_register_compressed("demo", dataDir, 3);
const db = turbolite_open(join(dataDir, "books.db"), "demo");
turbolite_exec(db, `
  CREATE TABLE books (
    id INTEGER PRIMARY KEY AUTOINCREMENT,
    title TEXT NOT NULL,
    year INTEGER
  )
`);

// ── HTTP server ─────────────────────────────────────────────────────

const server = http.createServer((req, res) => {
  if (req.method === "GET" && req.url === "/books") {
    const json = turbolite_query_json(db, "SELECT * FROM books ORDER BY year");
    res.writeHead(200, { "Content-Type": "application/json" });
    res.end(json);

  } else if (req.method === "POST" && req.url === "/books") {
    let body = "";
    req.on("data", (c) => (body += c));
    req.on("end", () => {
      const { title, year } = JSON.parse(body);
      turbolite_exec(db, `INSERT INTO books (title, year) VALUES ('${title}', ${year})`);
      const json = turbolite_query_json(db, "SELECT * FROM books ORDER BY id DESC LIMIT 1");
      res.writeHead(201, { "Content-Type": "application/json" });
      res.end(json);
    });

  } else {
    res.writeHead(404);
    res.end("Not found");
  }
});

process.on("SIGINT", () => {
  turbolite_close(db);
  rmSync(dataDir, { recursive: true, force: true });
  process.exit(0);
});

server.listen(3000, () => {
  console.log(`turbolite ${turbolite_version()} — listening on :3000`);
});

File: examples/c/basic.c
/*
 * turbolite example — C
 *
 * A sensor data logger that writes readings to compressed SQLite.
 *
 * Build & run:
 *   make lib-bundled header
 *   make example-c
 */

#include "../../turbolite.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

static void die(const char *msg) {
    const char *err = turbolite_last_error();
    fprintf(stderr, "FATAL: %s: %s\n", msg, err ? err : "unknown");
    exit(1);
}

static char tmpdir_buf[256];

int main(void) {
    /* Create data directory */
    snprintf(tmpdir_buf, sizeof(tmpdir_buf), "%s/turbolite-sensor-XXXXXX",
             getenv("TMPDIR") ? getenv("TMPDIR") : "/tmp");
    const char *data_dir = mkdtemp(tmpdir_buf);
    if (!data_dir) { perror("mkdtemp"); return 1; }

    printf("turbolite %s — sensor logger\n", turbolite_version());
    printf("Data dir: %s\n\n", data_dir);

    /* Register VFS and open database */
    if (turbolite_register_compressed("sensor", data_dir, 3) != 0)
        die("register VFS");

    char db_path[512];
    snprintf(db_path, sizeof(db_path), "%s/sensors.db", data_dir);

    void *db = turbolite_open(db_path, "sensor");
    if (!db) die("open database");

    turbolite_exec(db,
        "CREATE TABLE readings ("
        "  id INTEGER PRIMARY KEY AUTOINCREMENT,"
        "  sensor TEXT NOT NULL,"
        "  value REAL NOT NULL,"
        "  ts INTEGER NOT NULL"
        ")");

    /* Simulate logging sensor readings */
    srand((unsigned)time(NULL));
    for (int i = 0; i < 10; i++) {
        char sql[256];
        double temp = 20.0 + (rand() % 100) / 10.0;
        snprintf(sql, sizeof(sql),
            "INSERT INTO readings (sensor, value, ts) VALUES ('temp-01', %.1f, %ld)",
            temp, (long)time(NULL) + i);

        if (turbolite_exec(db, sql) != 0)
            die("insert reading");
    }
    printf("Logged 10 readings.\n");

    /* Query summary */
    char *json = turbolite_query_json(db,
        "SELECT sensor, COUNT(*) as count, "
        "ROUND(AVG(value), 1) as avg_temp, "
        "ROUND(MIN(value), 1) as min_temp, "
        "ROUND(MAX(value), 1) as max_temp "
        "FROM readings GROUP BY sensor");
    if (!json) die("query");

    printf("Summary: %s\n", json);
    turbolite_free_string(json);

    turbolite_close(db);
    return 0;
}
